plugin/trino-prometheus/src/main/java/io/trino/plugin/prometheus/expression/RewriteIn.java

```diff
@@ -0,0 +1,84 @@
+import io.trino.matching.Pattern;
+import io.trino.plugin.base.expression.ConnectorExpressionRule;
+import io.trino.spi.expression.Call;
+import io.trino.spi.expression.ConnectorExpression;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static io.trino.matching.Capture.newCapture;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argument;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.argumentCount;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.arguments;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.call;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.functionName;
+import static io.trino.plugin.base.expression.ConnectorExpressionPatterns.type;
+import static io.trino.plugin.prometheus.expression.RewriteHelpers.LABEL_MAP_CHECK_CORRECT_TYPE;
+import static io.trino.plugin.prometheus.expression.RewriteHelpers.extractLabelName;
+import static io.trino.plugin.prometheus.expression.RewriteHelpers.extractVarchar;
+import static io.trino.spi.expression.StandardFunctions.ARRAY_CONSTRUCTOR_FUNCTION_NAME;
+import static io.trino.spi.expression.StandardFunctions.IN_PREDICATE_FUNCTION_NAME;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+
+public class RewriteIn
+        implements ConnectorExpressionRule<Call, ImmutableList<LabelFilterExpression>>
+{
+    private static final Capture<Call> LABEL_NAME = newCapture();
+    private static final Capture<List<ConnectorExpression>> LABEL_VALUES = newCapture();
+
+    private static final Pattern<Call> PATTERN = call()
+            .with(functionName().equalTo(IN_PREDICATE_FUNCTION_NAME))
+            .with(type().equalTo(BOOLEAN))
+            .with(argumentCount().equalTo(2))
+            .with(argument(0).matching(call()
+                    .matching(LABEL_MAP_CHECK_CORRECT_TYPE)
+                    .capturedAs(LABEL_NAME)))
+            .with(argument(1).matching(call()
+                    .with(functionName().equalTo(ARRAY_CONSTRUCTOR_FUNCTION_NAME))
+                    .matching(call -> call.getArguments().stream().map(ConnectorExpression::getType).allMatch(RewriteHelpers::isCharOrStringType))
+                    .with(arguments().capturedAs(LABEL_VALUES))));
+
+    @Override
+    public Pattern<Call> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<ImmutableList<LabelFilterExpression>> rewrite(Call call, Captures captures, RewriteContext<ImmutableList<LabelFilterExpression>> context)
+    {
+        String labelName = extractLabelName(captures, LABEL_NAME);
+        List<ConnectorExpression> expressions = captures.get(LABEL_VALUES);
+
+        List<String> terms = new ArrayList<>(expressions.size());
+        for (ConnectorExpression expression : expressions) {
+            terms.add(extractVarchar(expression));
+        }
```

Comment:
nit: this section could be rewritten with Java Streams:
```
List<String> terms = expressions.stream()
    .map(this::extractVarchar)
    .collect(Collectors.toList());
 ```
 
 or even (if you consider inlining `expressions`
 ```
 List<String> terms = captures.get(LABEL_VALUES).stream()
    .map(this::extractVarchar)
    .collect(Collectors.toList());
```

Reply:
Great idea. Actually I think we can push it even further and even avoid creating the list:

```java
        String labelName = extractLabelName(captures, LABEL_NAME);
        List<ConnectorExpression> expressions = captures.get(LABEL_VALUES);

        return Optional.of(ImmutableList.of(new LabelFilterExpression(
                labelName,
                expressions.stream()
                        .map(RewriteHelpers::extractVarchar)
                        .collect(Collectors.joining("|", "(", ")")),
                true,
                false)));
```